### ISSUE TITLE
Port YouTube Description Generator Script

### ISSUE THREAD

**Issue by shiffman** [CodingTrain/thecodingtrain.com#205]
The old site had [a super convenient script to take a markdown page for a video and generate the YouTube description](https://github.com/CodingTrain/website-archive/blob/main/_scripts/generate-youtube-descriptions.js) for it. It was originally written by the amazing @duskvirkus!

I'd like to eventually port that script to work with the new site and (dare I say) maybe even eventually integrate with the YouTube API itself???

I know @klinegareth has been thinking about this and developing a CLI for the site as well. Let's discuss!

I was reminded of this when thinking about #203 and redirects and all the links from YouTube video descriptions that are now broken. 😬

**Comment by dipamsen** [comment]
Here is a draft version I created: https://gist.github.com/dipamsen/8100b65ec03262a38cf066cd6e67c837

**Comment by shiffman** [comment]
This is fantastic @dipamsen! I don't have time to review it right now but I will take a look very soon! I don't know the best way to integrate additional scripts but I could imagine something like the following being very useful. 

```console
npm run yt-desc /path/to/video
```

Or maybe we even make it an endpoint? @fdoflorenzano @runemadsen would adding a YouTube description generator script into this repo affect the website code / build process in any way that I might not be anticipating?

**Comment by fdoflorenzano** [comment]
@shiffman Can you explain what you expect from this script? I'm not sure why a YouTube descriptions script needs to live in this repo. It wouldn't create content for the actual website, right?

**Comment by runemadsen** [comment]
@shiffman @fdoflorenzano Jumping in here :) I remember us talking about this script earlier on, and as far as I remember, it's for simplifying the task of adding a video to the website and then to YouTube. So it just spits out the stuff needed for YouTube based on the website content. In that sense, I don't mind it being in this repo, and I think adding it as a script in `package.json` is a good idea. I propose adding it as a stand-alone script to `node-scripts`. It shouldn't affect the build process in any way. Happy to look through a PR if you tag me in it.

**Comment by shiffman** [comment]
Thanks everyone! @dipamsen do you want to bring this over in a branch here so we can start collaborating on it? This is a nice opportunity to add some features and improve this script. Tagging @KobeLiesenborgs to follow as well!

**Comment by dipamsen** [comment]
@shiffman Created a branch [here](https://github.com/CodingTrain/thecodingtrain.com/tree/description-script).

Running `npm run yt-desc` on this branch should spit out all the description files in a `_description` directory. (Running the script for a specific video is not supported, yet.)

-----

**Some Technical Notes/Remarks**
- Unlike the old repo, there is no way to know about a track/playlist just from the video file. For this the additional information in `content/tracks` is required. Before starting the generation, the script parses all the `tracks` information files.
- Also, there is no mention of the youtube playlist URL/`playlistId` in the content files. For now, the script has the hardcoded playlist IDs of Coding Challenges, Code! and NOC2. A solution would be to add a `playlistId` field to the track's json file. (this will have no role in the website itself, and some tracks won't have this field, eg: p5 tips and tricks). Or maybe the playlist URL doesn't need to be in the description at all?
- Earlier, only the p5.js Web Editor sketches were listed in the description. Should this be changed to have other code examples as well?

**Comment by shiffman** [comment]
Thank you @dipamsen! Just returning to taking a look at this today!  Here are some thoughts:

> Unlike the old repo, there is no way to know about the track/playlist just from the video file. For this the additional information in content/tracks is required. To achieve this, the script parses all the tracks information files before starting the generation.

One added complexity here is that a video can be in multiple tracks, I think we can largely ignore this for the YT video description and adding a playlistId for a "canonical" YouTube playlist to associate it with makes sense.

> Also, there is no mention of the youtube playlist URL/playlistId in the content files. For now, the script has the [hardcoded playlist IDs](https://github.com/CodingTrain/thecodingtrain.com/blob/3bd5c47f133b8d369003aa8cf22a5b041fc16439/node-scripts/yt-description.js#L169-L175) of Coding Challenges, Code! and NOC2. A solution would be to add a playlistId field to the track's json file. (this will have no role in the website itself, and some tracks won't have this field, eg: p5 tips and tricks). Or maybe the playlist URL doesn't need to be in the description at all?

Ah, yes, I think for the purposes of syncing data with YouTube description I would like to add a `playlistId` field that is ignored for the website but we use for our own data purposes. @runemadsen @fdoflorenzano is there any issue / unintended consequence with us doing this?

> Earlier, only the p5.js Web Editor sketches were listed in the description. Should this be changed to have other code examples as well? (Standalone GitHub repos for challenges should be added, if any!)

I'm not sure about this. In theory, if we have the link to the webpage that's where someone can go to find all the code, but it also doesn't hurt to include extra links in the YT description. What do you think? As an example this Ulam spiral challenge includes a bunch of code links and I think it looks nice! (https://www.youtube.com/watch?v=a35KWEjRvc0)

Please open a [WIP] pull request from this branch if you like!

**Comment by shiffman** [comment]
@dipamsen Would you like to keep this thread open or shall we close and address anything additional as it comes up separately! The script is working amazingly well!

**Comment by dipamsen** [comment]
I think we can close this for now! I would like some confirmation if the `playlistId` field in tracks can be added so that I can update it to dynamically get playlist links. I am guessing that content tests will have to be updated for that to work.

**Comment by lobertrand** [comment]
Hi! Maybe this script should take into account the special case of multi-part videos. There's a new `"parts"` section in the JSON file dedicated to them. It follows the format described in this comment : https://github.com/CodingTrain/thecodingtrain.com/issues/420#issuecomment-1218529904 .

**Comment by shiffman** [comment]
@dipamsen what do you think about adding something for the multi-part challenges?

**Comment by dipamsen** [comment]
I will work on updating the script to work will multi-part challenges. The structure of the script might have to be changed since currently each index.json is treated as one video.

**Comment by shiffman** [comment]
@dipamsen once we finalize the multi-part challenge feature I think we can close this issue and open individual ones to track enhancements or bugs!

**Comment by shiffman** [comment]
Well, here are two feature requests curious for your thoughts @dipamsen. One hard and one easy.

1. I would be really convenient if I could somehow use the website URL of the video and/or else the full path for the index.json file. This is how I find myself actually wanting to use the script.

2. I think the URL that goes to thecodingtrain.com page in the description should have something to indicate that's where you find the code. Maybe just:

> Welcome aboard! This is the introductory video to start programming JavaScript with p5.js for total beginners. Code: https://thecodingtrain.com/tracks/code-programming-with-p5-js/code/1-intro/1-intro

or

> Welcome aboard! This is the introductory video to start programming JavaScript with p5.js for total beginners. Find the code and see the showcase! https://thecodingtrain.com/tracks/code-programming-with-p5-js/code/1-intro/1-intro

or

> Welcome aboard! This is the introductory video to start programming JavaScript with p5.js for total beginners. 💻 https://thecodingtrain.com/tracks/code-programming-with-p5-js/code/1-intro/1-intro

Thoughts?

**Comment by dipamsen** [comment]
> I would be really convenient if I could somehow use the website URL of the video and/or else the full path for the index.json file. This is how I find myself actually wanting to use the script.

The script can be updated to use anyone of these:
- complete website url
- path to file
- even just YT videoId! (since *all* video files are parsed prior to generating the description)

For clarification, you would like to use the website URL in the CLI to generate the description:
```
npm run yt-desc https://thecodingtrain.com/challenges/171-wave-function-collapse
``` 
Is this correct?

> I think the URL that goes to thecodingtrain.com page in the description should have something to indicate that's where you find the code. 

I am down for whatever! I think both `Code:` or `Find the code and showcase` works.

**Comment by shiffman** [comment]
> 
> For the sake of clarification, you would like to use the website URL in the CLI to generate the description:
> 
> ```
> npm run yt-desc https://thecodingtrain.com/challenges/171-wave-function-collapse
> ```
> 
> Is this correct?

Yes!

> > I think the URL that goes to thecodingtrain.com page in the description should have something to indicate that's where you find the code.
> 
> I agree! I am down for whatever. I think both `Code:` or `Find the code and showcase` works.

Great! Let's try "Code:" since this is 9 times out of 10 what the viewer is looking for.

**Comment by shiffman** [comment]
@dipamsen what do you think about adding the language and topics keywords at the very bottom with hashtags? (any spaces would have to be removed) This is low priority but might help with searchability on YT?

**Comment by shiffman** [comment]
Closing this!!! 🚀
